pkg/webhooks/kafkauser_validator.go

```diff
@@ -0,0 +1,117 @@
+	"github.com/banzaicloud/koperator/pkg/util"
+)
+
+type KafkaUserValidator struct {
+	Client client.Client
+	Log    logr.Logger
+}
+
+// ValidateCreate validates if the user-provided KafkaUser specifications valid for creation
+func (validator KafkaUserValidator) ValidateCreate(ctx context.Context, obj runtime.Object) error {
+	return validator.validate(ctx, obj)
+}
+
+// ValidateUpdate validates if the user-provided KafkaUser specifications valid for update
+func (validator KafkaUserValidator) ValidateUpdate(ctx context.Context, _, newObj runtime.Object) error {
+	return validator.validate(ctx, newObj)
+}
+
+// ValidateDeleet validates if the user-selctd KafkaUser specifications valid for deletion
+func (validator KafkaUserValidator) ValidateDelete(_ context.Context, _ runtime.Object) error {
+	// nothing to validate on deletion
+	return nil
+}
+
+func (validator KafkaUserValidator) validate(ctx context.Context, obj runtime.Object) error {
+	var (
+		kafkaUser = obj.(*banzaicloudv1alpha1.KafkaUser)
+		log       = validator.Log.WithValues("name", kafkaUser.GetName(), "namespace", kafkaUser.GetNamespace())
+	)
+
+	fieldErrs, err := validator.validateKafkaUser(ctx, kafkaUser)
+	if err != nil {
+		log.Error(err, errorDuringValidationMsg)
+		return apierrors.NewInternalError(errors.WithMessage(err, errorDuringValidationMsg))
+	}
+
+	if len(fieldErrs) == 0 {
+		return nil
+	}
+
+	log.Info("rejected", "invalid field(s)", fieldErrs.ToAggregate().Error())
+
+	return apierrors.NewInvalid(kafkaUser.GetObjectKind().GroupVersionKind().GroupKind(), kafkaUser.Name, fieldErrs)
+}
+
+func (validator KafkaUserValidator) validateKafkaUser(ctx context.Context, kafkaUser *banzaicloudv1alpha1.KafkaUser) (field.ErrorList, error) {
+	var (
+		err     error
+		allErrs field.ErrorList
+		logMsg  string
+
+		kafkaCluster     *v1beta1.KafkaCluster
+		clusterNamespace = util.GetClusterRefNamespace(kafkaUser.GetNamespace(), kafkaUser.Spec.ClusterRef)
+		clusterName      = kafkaUser.Spec.ClusterRef.Name
+	)
+
+	if kafkaCluster, err = k8sutil.LookupKafkaCluster(ctx, validator.Client, clusterName, clusterNamespace); err != nil {
+		if apierrors.IsNotFound(err) {
+			logMsg = fmt.Sprintf("KafkaCluster CR '%s' in the namespace '%s' does not exist", clusterName, clusterNamespace)
+			allErrs = append(allErrs, field.Invalid(field.NewPath("spec").Child("clusterRef"), kafkaUser.Spec.ClusterRef, logMsg))
+			return allErrs, nil
+		}
+
+		return nil, err
+	}
+
+	if k8sutil.IsMarkedForDeletion(kafkaCluster.ObjectMeta) {
+		return nil, errors.New("referenced KafkaCluster CR is being deleted")
+	}
+
+	if util.ObjectManagedByClusterRegistry(kafkaCluster) {
+		// referencing remote Kafka clusters is not allowed
+		logMsg = fmt.Sprintf("KafkaCluster CR '%s' in the namespace '%s' is a remote resource", clusterName, clusterNamespace)
+		allErrs = append(allErrs, field.Invalid(field.NewPath("spec").Child("clusterRef"), kafkaUser.Spec.ClusterRef, logMsg))
+	}
```

Comment:
**Question**: Won't this cause issues with clreg replicated KafkaUsers from remote K8s clusters and Kafka clusters?
I didn't see any guard against validating KafkaUsers replicated by clreg.

Reply:
Hmmm yea you are right, we probably shouldn't have this restriction here